streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskManager.java

```diff
@@ -59,11 +60,17 @@
 
     private final Admin adminClient;
     private DeleteRecordsResult deleteRecordsResult;
+    private boolean restoreConsumerAssignedStandbys = false;
```

Comment:
Related to my other comments: if we assume that the same thread never have active and standby of the same task at the same time then we can simplify this a bit: first we can check inside `onAssignment` to make sure it is indeed the case, and if not treat it as a fatal error. Then we can be ensured that the partitions do not have any overlaps and we can just "blindly" remove partitions as we remove standbys since if they are not assigned yet, the removal would be safely a no-op. The only usage of this would be line 124 below.

Reply:
Not sure I follow, why does it help us to check this in `onAssignment`? We should be confident there is no overlap (unless there is a bug, might be good to verify somewhere I suppose) -- `restoreConsumerAssignedStandbys` is meant as a flag whether it's assigned standby OR active, ie this would be false if it is assigned any active.

Reply:
Technically it should always be safe to remove any revoked standby or actives since its a no-op if not assigned them anyway. This is what the PR was doing originally, which was "accidentally correct" since I wasn't aware it was only assigned one or the other.

Reply:
My only motivation is to simplify the usage of the flag a bit :) I feel that some conditions on this flag is not necessary.

Reply:
I was thinking of it more as a slight optimization to avoid making a copy of the assignment and re-assigning, but maybe it's better to play it safe & decrease complexity by just always removing the partitions

Reply:
Moved the removing changelogs from restore consumer into a separate method. I left in the check for whether its assigned/removing standbys or active but (I think) I simplified it a bit -- happy to take it out if you feel strongly though (note we can't get rid of it entirely as we need to check eg whether to unsubscribe all if its reading standbys and there are new restoring tasks

Reply:
@ableegoldman this part of the code seems not changing to me?

Reply:
Sorry, what seems not changed?
`restoreConsumerAssigneStandbys` is now used in only two places, first to determine whether it should unsubscribe all in case there are new restoring tasks, and second in `removeChangelogsFromRestoreConsumer` where we skip removing partitions if they weren't what it was assigned to (to avoid overhead of copying assignment)

Reply:
Ah I think I was looking at an old commit :P LGTM now.